### PR TITLE
fix: proper errors for instance launch problems

### PR DESCRIPTION
**Pull request by aaronkvanmeerten**
count instance success when launching
move generic code from oracle instance launcher
handle new instances in cloud manager, not oracle manager
count error when less instances launch than attempted

### CODE REVIEW

**Comment by aaronkvanmeerten** [comment]
This moves generic instance handling code out of the oracle instance manager, leaving the oracle manager to handle only oracle-specific code
This also adds counts errors when attempting to launch instances and failing

File: src/cloud_manager.ts [hunk 1]
```diff
@@ -56,20 +55,53 @@ export default class CloudManager {
...
+                ctx.logger.error(`Cloud type not supported: ${group.cloud}`);
+                return 0;
         }
-        return true;
+
+        let scaleUpCount = 0;
+        scaleUpResult.forEach(async (instanceId) => {
+            if (instanceId) {
+                scaleUpCount++;
+                if (!this.isDryRun && instanceId !== true) {
```
Comment (oanaianc) [inline src/cloud_manager.ts:82]: We should also check for instanceId !== false, because it can be false

File: src/instance_launcher.ts [hunk 2]
```diff
@@ -80,18 +80,46 @@ export default class InstanceLauncher {
...
-                // increment launched instance stats for the group
-                instancesLaunchedCounter.inc({ group: group.name }, actualScaleUpQuantity);
+                    // check if scale up count didn't meet requested quantity, error if so
+                    if (scaleUpCount != actualScaleUpQuantity) {
+                        ctx.logger.error(
+                            `[Launcher] Scaling failed to launch requested new instances for group ${groupName} with ${count} instances.`,
+                            { scaleUpRequested: actualScaleUpQuantity, scaleUpActual: scaleUpCount },
+                        );
+                        instanceErrorsCounter.inc({ group: group.name });
+                        throw new Error(
```
Comment (oanaianc) [inline src/instance_launcher.ts:110]: This error is caught locally and so the instanceErrorsCounter will be incremented twice; we also increment the instanceErrorsCounter on catch (err)

File: src/cloud_manager.ts [hunk 3]
```diff
@@ -56,20 +55,53 @@ export default class CloudManager {
...
+                );
+                break;
+            default:
+                ctx.logger.error(`Cloud type not supported: ${group.cloud}`);
+                return 0;
         }
-        return true;
+
+        let scaleUpCount = 0;
+        scaleUpResult.forEach(async (instanceId) => {
```
Comment (oanaianc) [inline src/cloud_manager.ts:79]: We might want to replace forEach with for/of loop. Initially I saw it as an IntelliJ suggestion then I read more and it doesn't seem appropriate to use await within a forEach() callback

Comment (aaronkvanmeerten) [reply src/cloud_manager.ts]: Agreed, fixed.

File: src/instance_launcher.ts (hunk 2 above)
Comment (aaronkvanmeerten) [reply src/instance_launcher.ts:109]: Agreed, good catch.

File: src/cloud_manager.ts (hunk 1 above)
Comment (aaronkvanmeerten) [reply src/cloud_manager.ts]: It cannot be false as we check for instanceId above

**Review by oanaianc — APPROVED** [review]

**Pull request merged**